File: src/main/java/com/amberclient/modules/combat/AutoClicker.java
package com.amberclient.modules.combat;

import com.amberclient.utils.module.ConfigurableModule;
import com.amberclient.utils.module.Module;
import com.amberclient.utils.module.ModuleSettings;
import net.minecraft.client.MinecraftClient;
import net.minecraft.entity.Entity;
import net.minecraft.entity.attribute.EntityAttributes;
import net.minecraft.item.ItemStack;
import net.minecraft.text.Text;
import net.minecraft.util.Hand;
import net.minecraft.util.hit.EntityHitResult;
import net.minecraft.util.hit.HitResult;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.ArrayList;
import java.util.List;

public class AutoClicker extends Module implements ConfigurableModule {
    public static final String MOD_ID = "amberclient-autoclicker";
    public static final Logger LOGGER = LogManager.getLogger(MOD_ID);

    private final List<ModuleSettings> settings;
    private final ModuleSettings clickDelay;
    private final ModuleSettings useWeaponSpeed;
    private long lastClickTime;

    public AutoClicker() {
        super("AutoClicker", "Auto clicks when aiming at an entity", "Combat");

        useWeaponSpeed = new ModuleSettings("Use Weapon Speed", "Use the attack delay of the held item", true);
        clickDelay = new ModuleSettings("Default Click Delay", "Delay between clicks in ms (when weapon speed is disabled)", 100.0, 50.0, 500.0, 10.0);
        lastClickTime = 0;

        settings = new ArrayList<>();
        settings.add(useWeaponSpeed);
        settings.add(clickDelay);
    }

    private double getWeaponAttackDelay() {
        MinecraftClient client = MinecraftClient.getInstance();
        if (client.player == null) return 100.0;

        ItemStack heldItem = client.player.getMainHandStack();
        if (heldItem.isEmpty()) {
            double baseAttackSpeed = client.player.getAttributeValue(EntityAttributes.ATTACK_SPEED);
            return 1000.0 / baseAttackSpeed;
        }

        double attackSpeed = client.player.getAttributeValue(EntityAttributes.ATTACK_SPEED);

        return 1000.0 / attackSpeed;
    }

    @Override
    public void onTick() {
        MinecraftClient client = MinecraftClient.getInstance();
        if (client.player == null || client.world == null) return;

        HitResult hitResult = client.crosshairTarget;
        if (hitResult != null && hitResult.getType() == HitResult.Type.ENTITY) {
            EntityHitResult entityHitResult = (EntityHitResult) hitResult;
            Entity target = entityHitResult.getEntity();
            long currentTime = System.currentTimeMillis();

            double delayToUse;
            if (useWeaponSpeed.getBooleanValue()) {
                delayToUse = getWeaponAttackDelay();
            } else {
                delayToUse = clickDelay.getDoubleValue();
            }

            if (currentTime - lastClickTime >= delayToUse) {
                client.interactionManager.attackEntity(client.player, target);
                client.player.swingHand(Hand.MAIN_HAND);
                lastClickTime = currentTime;
            }
        }
    }

    @Override
    public List<ModuleSettings> getSettings() { return settings; }

    @Override
    public void onSettingChanged(ModuleSettings setting) {
        MinecraftClient client = MinecraftClient.getInstance();
        if (client.player == null) return;

        if (setting == clickDelay) {
            client.player.sendMessage(
                    Text.literal("§6AutoClicker settings updated: Click Delay=" + clickDelay.getDoubleValue() + " ms"),
                    true);
            LOGGER.info("AutoClicker settings updated: Click Delay={} ms", clickDelay.getDoubleValue());
        } else if (setting == useWeaponSpeed) {
            String status = useWeaponSpeed.getBooleanValue() ? "enabled" : "disabled";
            client.player.sendMessage(
                    Text.literal("§6AutoClicker settings updated: Use Weapon Speed=" + status),
                    true);
            LOGGER.info("AutoClicker settings updated: Use Weapon Speed={}", status);
        }
    }
}

File: src/main/java/com/amberclient/modules/combat/AutoClicker.kt
package com.amberclient.modules.combat

import com.amberclient.utils.module.ConfigurableModule
import com.amberclient.utils.module.Module
import com.amberclient.utils.module.ModuleSettings
import net.minecraft.client.MinecraftClient
import net.minecraft.entity.Entity
import net.minecraft.entity.attribute.EntityAttributes
import net.minecraft.item.ItemStack
import net.minecraft.text.Text
import net.minecraft.util.Hand
import net.minecraft.util.hit.EntityHitResult
import net.minecraft.util.hit.HitResult
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger

class AutoClicker : Module("AutoClicker", "Auto clicks when aiming at an entity", "Combat"), ConfigurableModule {

    companion object {
        const val MOD_ID = "amberclient-autoclicker"
        val LOGGER: Logger = LogManager.getLogger(MOD_ID)
    }

    private val useWeaponSpeed: ModuleSettings
    private val clickDelay: ModuleSettings
    private var lastClickTime: Long = 0
    private val settings: MutableList<ModuleSettings>

    init {
        useWeaponSpeed = ModuleSettings("Use Weapon Speed", "Use the attack delay of the held item", true)
        clickDelay = ModuleSettings("Default Click Delay", "Delay between clicks in ms (when weapon speed is disabled)", 100.0, 50.0, 500.0, 10.0)

        settings = mutableListOf<ModuleSettings>().apply {
            add(useWeaponSpeed)
            add(clickDelay)
        }
    }

    private fun getWeaponAttackDelay(): Double {
        val client = MinecraftClient.getInstance()
        val player = client.player ?: return 100.0

        val heldItem: ItemStack = player.mainHandStack
        if (heldItem.isEmpty) {
            val baseAttackSpeed = player.getAttributeValue(EntityAttributes.ATTACK_SPEED)
            return 1000.0 / baseAttackSpeed
        }

        val attackSpeed = player.getAttributeValue(EntityAttributes.ATTACK_SPEED)
        return 1000.0 / attackSpeed
    }

    override fun onTick() {
        val client = MinecraftClient.getInstance()
        val player = client.player ?: return
        val world = client.world ?: return

        val hitResult = client.crosshairTarget
        if (hitResult != null && hitResult.type == HitResult.Type.ENTITY) {
            val entityHitResult = hitResult as EntityHitResult
            val target: Entity = entityHitResult.entity
            val currentTime = System.currentTimeMillis()

            val delayToUse = if (useWeaponSpeed.booleanValue) {
                getWeaponAttackDelay()
            } else {
                clickDelay.doubleValue
            }

            if (currentTime - lastClickTime >= delayToUse) {
                client.interactionManager?.let { interactionManager ->
                    interactionManager.attackEntity(player, target)
                    player.swingHand(Hand.MAIN_HAND)
                    lastClickTime = currentTime
                }
            }
        }
    }

    override fun getSettings(): List<ModuleSettings> = settings

    override fun onSettingChanged(setting: ModuleSettings) {
        val client = MinecraftClient.getInstance()
        val player = client.player ?: return

        when (setting) {
            clickDelay -> {
                player.sendMessage(
                    Text.literal("§6AutoClicker settings updated: Click Delay=${clickDelay.doubleValue} ms"),
                    true
                )
                LOGGER.info("AutoClicker settings updated: Click Delay={} ms", clickDelay.doubleValue)
            }
            useWeaponSpeed -> {
                val status = if (useWeaponSpeed.booleanValue) "enabled" else "disabled"
                player.sendMessage(
                    Text.literal("§6AutoClicker settings updated: Use Weapon Speed=$status"),
                    true
                )
                LOGGER.info("AutoClicker settings updated: Use Weapon Speed={}", status)
            }
        }
    }
}